enhancements/installer/aws-permissions-check-bypass.md

```diff
@@ -0,0 +1,318 @@
+---
+title: aws-permissions-check-bypass
+authors:
+  - "@joelddiaz"
+reviewers:
+  - "@dgoodwin"
+  - "@abhinavdahiya"
+  - "@jeremyeder"
+approvers:
+  - TBD
+creation-date: 2020-05-07
+last-updated: 2020-05-15
+status: provisional
+see-also:
+  - ""  
+replaces:
+  - ""
+superseded-by:
+  - ""
+---
+
+# AWS Permissions Check Bypass
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Open Questions [optional]
+
+## Summary
+
+Running the OpenShift installer in AWS accounts using Service Control Policies
+(SCP) can result in errors indicating that the credentials provided do not have
+sufficient permissions when in reality they do. This is a limitation with the
+AWS IAM permissions simulation API.  Additionally, even without these
+limitations it is also true that the AWS policy language is sophisticated enough
+to restrict permissions in a way where is is difficult to perform acurate policy
+simulations (eg restricting permissions based on VPC or source IP).
+
+## Motivation
+
+Presently, when installing OpenShift on AWS, the credentials provided to the
+installer are queried/checked for appropriate permissions to verify that:
+1. Can be used to complete an installation (the installer set of required permissions)
+1. The in-cluster cloud-credential-operator (CCO) can successfully operate in
+   either 'mint' or 'passthrough' mode.
+If any of these conditions are not met, the install will fail and explain which
+permissions were deemed to be missing.
+
+When an AWS account is configured with Service Control Policies (SCP) the
+permissions checking/simulation APIs can provide incorrect results depending on
+the contents of the SCPs. Service Control Policies are typically used to deny
+certain API calls unless a condition (or set of conditions) is met.  For
```

Comment:
+1 i will add some text to go into the details.

Reply:
@jharrington22 I'll document this here for posterity. I went to try out the SCP setup without anything involving region conditions, and the policy simulation still seems unusable. Specifically, I denied the API call "rds:DescribeDBInstance" to any user named "jdiaz", and the permissions simulation says "Allowed" when in reality it is not allowed:

```
[jdiaz@minigoomba ~]$ aws iam simulate-principal-policy --policy-source-arn arn:aws:iam::ACCOUNT_NUMBER:user/jdiaz --action-names '["rds:DescribeDBInstances"]' --context-entries ContextKeyName=aws:RequestedRegion,ContextKeyType=string,ContextKeyValues=us-east-1 | jq -r .EvaluationResults[0].EvalDecision
allowed
[jdiaz@minigoomba ~]$ aws rds describe-db-instances --region us-east-1

An error occurred (AccessDenied) when calling the DescribeDBInstances operation: User: arn:aws:iam::ACCOUNT_NUMBER:user/jdiaz is not authorized to perform: rds:DescribeDBInstances with an explicit deny
```
So while this is a false positive situation that at least would allow the installation to proceed before failing, the whole permissions simulation just can't be trusted with AWS SCPs influencing things.

I'll add some text covering the false positives and false negatives cases.

Reply:
@joelddiaz interesting seems that conditions on SCP policies aren't supported. I'll bring this specific example up with AWS as i'm not sure they are aware that isn't not just related to regions.